packages/gatsby-source-contentful/src/backreferences.js

```diff
@@ -0,0 +1,142 @@
+import { getDataStore } from "gatsby/dist/datastore"
+import { untilNextEventLoopTick } from "./utils"
+
+// Array of all existing Contentful nodes. Make it global and incrementally update it because it's hella slow to recreate this on every data update for large sites.
+export const existingNodes = new Map()
+
+let allNodesLoopCount = 0
+
+// "is" === object so it can be overridden by tests
+export const is = {
+  firstSourceNodesCallOfCurrentNodeProcess: true,
+}
+
+export async function getExistingCachedNodes({
+  actions,
+  getNode,
+  pluginConfig,
+}) {
+  const { touchNode } = actions
+
+  const needToTouchNodes =
+    !hasFeature(`stateful-source-nodes`) &&
+    is.firstSourceNodesCallOfCurrentNodeProcess
+
+  if (existingNodes.size === 0) {
+    memoryNodeCountsBySysType.Asset = 0
+    memoryNodeCountsBySysType.Entry = 0
+
+    const dataStore = getDataStore()
+    const allNodeTypeNames = Array.from(dataStore.getTypes())
+
+    for (const typeName of allNodeTypeNames) {
+      const typeNodes = dataStore.iterateNodesByType(typeName)
+
+      const firstNodeOfType = Array.from(typeNodes.slice(0, 1))[0]
+
+      if (
+        !firstNodeOfType ||
+        firstNodeOfType.internal.owner !== `gatsby-source-contentful`
+      ) {
+        continue
+      }
+
+      for (const node of typeNodes) {
+        if (needToTouchNodes) {
+          touchNode(node)
+
+          if (node?.fields?.includes(`localFile`)) {
+            // Prevent GraphQL type inference from crashing on this property
+            const fullNode = getNode(node.id)
+            const localFileNode = getNode(fullNode.fields.localFile)
+            touchNode(localFileNode)
+          }
+        }
+
+        if (++allNodesLoopCount % 5000 === 0) {
+          // dont block the event loop
+          await untilNextEventLoopTick()
+        }
+
+        if (
+          pluginConfig.get(`enableTags`) &&
+          node.internal.type === `ContentfulTag`
+        ) {
+          continue
+        }
+
+        addNodeToExistingNodesCache(node)
+      }
+
+      // dont block the event loop
+      await untilNextEventLoopTick()
+    }
+  }
+
+  is.firstSourceNodesCallOfCurrentNodeProcess = false
+
+  return {
+    existingNodes,
+    memoryNodeCountsBySysType,
+  }
+}
+
+const memoryNodeCountsBySysType = {
+  Asset: 0,
+  Entry: 0,
+}
+
+// store only the fields we need to compare to reduce memory usage. if a node is updated we'll use getNode to grab the whole node before updating it
+export function addNodeToExistingNodesCache(node) {
+  if (
+    // no sys.type for ContentfulTag nodes
+    node?.sys?.type &&
+    node.sys.type in memoryNodeCountsBySysType &&
+    !existingNodes.has(node.id)
+  ) {
+    memoryNodeCountsBySysType[node.sys.type] ||= 0
+    memoryNodeCountsBySysType[node.sys.type]++
+  }
+
+  const cacheNode = {
+    id: node.id,
+    contentful_id: node.contentful_id,
+    sys: {
+      // no sys.type for ContentfulTag nodes
+      type: node?.sys?.type,
+    },
```

Comment:
is `sys` undefined or is `sys.type` undefined? I just wonder wether we should have `sys` at all, if only available field (`type`) would be `undefined`

Reply:
this is why it's undefined

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-contentful/src/source-nodes.js#L530-L538

Reply:
https://github.com/gatsbyjs/gatsby/pull/37910/commits/6a441009f85e3e15d1366a85430b317d69722d0d